packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
 			);
 			if (messagesWithPendingState.length > 0) {
-				messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
-					const msg: MessageWithContext = {
-						message,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-						localOpMetadata,
-					};
-					this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
-				});
+				this.batchBegin(messagesWithPendingState[0]?.message);
+
+				let error: any;
+				try {
+					messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
+						const msg: MessageWithContext = {
+							message,
+							local,
+							isRuntimeMessage: true,
+							savedOp,
+							localOpMetadata,
+						};
+						this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
+					});
+				} catch (e) {
+					error = e;
+					throw e;
+				} finally {
+					this.batchEnd(
+						error,
+						messagesWithPendingState[messagesWithPendingState.length - 1]?.message,
+					);
+				}
 			} else {
 				this.ensureNoDataModelChanges(() => this.processEmptyBatch(inboundBatch, local));
 			}
 		} else {
-			// Check if message.type is one of values in ContainerMessageType
-			// eslint-disable-next-line import/no-deprecated
-			if (isRuntimeMessage(messageCopy)) {
-				// Legacy op received
-				this.ensureNoDataModelChanges(() =>
-					this.processRuntimeMessage({
-						message: messageCopy as InboundSequencedContainerRuntimeMessage,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-					}),
-				);
-			} else {
-				// A non container runtime message (like other system ops - join, ack, leave, nack etc.)
-				this.ensureNoDataModelChanges(() =>
-					this.observeNonRuntimeMessage({
-						message: messageCopy as InboundSequencedNonContainerRuntimeMessage,
-						local,
-						isRuntimeMessage: false,
-						savedOp,
-					}),
-				);
+			this.batchBegin(messageCopy);
+			let error: any;
+			try {
+				// Check if message.type is one of values in ContainerMessageType
+				// eslint-disable-next-line import/no-deprecated
+				if (isRuntimeMessage(messageCopy)) {
+					// Legacy op received
+					this.ensureNoDataModelChanges(() =>
+						this.processRuntimeMessage({
+							message: messageCopy as InboundSequencedContainerRuntimeMessage,
+							local,
+							isRuntimeMessage: true,
+							savedOp,
+						}),
+					);
+				} else {
+					// A non container runtime message (like other system ops - join, ack, leave, nack etc.)
+					this.ensureNoDataModelChanges(() =>
+						this.observeNonRuntimeMessage({
+							message: messageCopy as InboundSequencedNonContainerRuntimeMessage,
+							local,
+							isRuntimeMessage: false,
+							savedOp,
+						}),
+					);
+				}
+			} catch (e) {
+				error = e;
+				throw e;
+			} finally {
+				this.batchEnd(error, messageCopy);
 			}
 		}
 	}
 
+	private batchBegin(message?: ISequencedDocumentMessage) {
+		assert(message !== undefined, "Batch must have a message");
+		this.deltaScheduler.batchBegin(message);
```

Comment:
What if the deltaScheduler listen for the "batchBegin" and "batchEnd" events?

As it is, the DeltaScheduler just sits there and does what it's told, and ContainerRuntime is the one who "knows" that it needs the DeltaScheduler to do something.  If we flip it, then the ContainerRuntime has less knowledge/opinion, and the DeltaScheduler is more independent, handling the events whenever they come.

I lean towards switching to event-based, it feels more modular - the ContainerRuntime doesn't bake in knowledge about the DeltaScheduler besides creating it.  We don't even need the class field anymore (but note that DeltaScheduler will need to listen for "dispose" event and dispose itself to release the event listeners)

Reply:
That sounds fine to me. However, I do not want to change that behavior in this PR. It can be done as a follow up. It will lead to other behavior change where this will be called for an empty batch which isn't a big deal but would be nice to do it separately.